src/qei.rs

```diff
@@ -0,0 +1,445 @@
+    feature = "stm32f437",
+    feature = "stm32f439",
+    feature = "stm32f446",
+    feature = "stm32f469",
+    feature = "stm32f479"
+))]
+impl PinC2<TIM2> for PA1<Alternate<AF1>> {}
+
+#[cfg(any(
+    feature = "stm32f401",
+    feature = "stm32f405",
+    feature = "stm32f407",
+    feature = "stm32f410",
+    feature = "stm32f411",
+    feature = "stm32f412",
+    feature = "stm32f413",
+    feature = "stm32f415",
+    feature = "stm32f417",
+    feature = "stm32f423",
+    feature = "stm32f427",
+    feature = "stm32f429",
+    feature = "stm32f437",
+    feature = "stm32f439",
+    feature = "stm32f446",
+    feature = "stm32f469",
+    feature = "stm32f479"
+))]
+impl PinC2<TIM2> for PB3<Alternate<AF1>> {}
+
+impl PinC1<TIM3> for PA6<Alternate<AF2>> {}
+
+impl PinC2<TIM3> for PA7<Alternate<AF2>> {}
+
+impl PinC1<TIM3> for PB4<Alternate<AF2>> {}
+
+impl PinC2<TIM3> for PB5<Alternate<AF2>> {}
+
+impl PinC1<TIM3> for PC6<Alternate<AF2>> {}
+
+impl PinC2<TIM3> for PC7<Alternate<AF2>> {}
+
+impl PinC1<TIM4> for PB6<Alternate<AF2>> {}
+
+impl PinC2<TIM4> for PB7<Alternate<AF2>> {}
+
+impl PinC1<TIM4> for PD12<Alternate<AF2>> {}
+
+impl PinC2<TIM4> for PD13<Alternate<AF2>> {}
+
+impl PinC1<TIM5> for PA0<Alternate<AF2>> {}
+
+impl PinC2<TIM5> for PA1<Alternate<AF2>> {}
+
+impl PinC1<TIM5> for PH10<Alternate<AF2>> {}
+
+impl PinC2<TIM5> for PH11<Alternate<AF2>> {}
+
+impl PinC1<TIM8> for PC6<Alternate<AF3>> {}
+
+impl PinC2<TIM8> for PC7<Alternate<AF3>> {}
+
+impl PinC1<TIM8> for PI5<Alternate<AF3>> {}
+
+impl PinC2<TIM8> for PI6<Alternate<AF3>> {}
+
+pub struct Qei<TIM, PINS> {
+    tim: TIM,
+    pins: PINS,
+}
+
+macro_rules! hal {
+    ($($TIM:ident: ($tim:ident, $timXen:ident, $timXrst:ident, $apbenr:ident, $apbrstr:ident, $bits:ident),)+) => {
+        $(
+            impl<PINS> Qei<$TIM, PINS> {
+                pub fn $tim(tim: $TIM, pins: PINS) -> Self {
+                    let rcc = unsafe { &(*RCC::ptr()) };
+                    // enable and reset peripheral to a clean slate state
+                    rcc.$apbenr.modify(|_, w| w.$timXen().set_bit());
+                    rcc.$apbrstr.modify(|_, w| w.$timXrst().set_bit());
+                    rcc.$apbrstr.modify(|_, w| w.$timXrst().clear_bit());
+
+                    // Configure TxC1 and TxC2 as captures
+                    tim.ccmr1_output
+                        .write(|w| unsafe { w.bits({ (0b01 << 0) | (0b01 << 8) }) });
+
+                    // enable and configure to capture on rising edge
+                    tim.ccer.write(|w| {
+                        w.cc1e()
+                            .set_bit()
+                            .cc1p()
+                            .clear_bit()
+                            .cc2e()
+                            .set_bit()
+                            .cc2p()
+                            .clear_bit()
+                    });
+
+                    // configure as quadrature encoder
+                    #[allow(unused_unsafe)]
```

Comment:
Huh? Unused unsafe?

Reply:
It's not the first time I hit this bug in rustc.

Without `unsafe {}`:
```
error[E0133]: call to unsafe function is unsafe and requires unsafe function or block
   --> src/qei.rs:809:40
    |
809 |                       tim.smcr.write(|w| w.sms().bits(3));
    |                                          ^^^^^^^^^^^^^^^ call to unsafe function
...
862 | / hal! {
863 | |     TIM1: (tim1, tim1en, tim1rst, apb2enr, apb2rstr, u16),
864 | |     TIM5: (tim5, tim5en, tim5rst, apb1enr, apb1rstr, u32),
865 | | }
    | |_- in this macro invocation
    |
    = note: consult the function's documentation for information on how to avoid undefined behavior
```

Reply:
It most likely is unsafe on one variant of stm32f4 and safe on another, or unsafe for TIM1 and safe for TIM5 for example...

Reply:
Yep, that's exactly the case. I've added a comment.